src/backend/renderer.js

```diff
@@ -832,109 +888,144 @@ export function attach(
-          } else {
-            mountFiber(
-              nextChild,
-              shouldEnqueueUpdate ? nextFiber : parentFiber
-            );
-            if (!hasChildOrderChanged) {
-              hasChildOrderChanged = true;
-            }
+    const shouldInclude = !shouldFilterFiber(nextFiber);
+    let hasChildOrderChanged = false;
+    if (nextFiber.child !== prevFiber.child) {
+      // If the first child is different, we need to traverse them.
+      // Each next child will be either a new child (mount) or an alternate (update).
+      let nextChild = nextFiber.child;
+      let prevChildAtSameIndex = prevFiber.child;
+      while (nextChild) {
+        // We already know children will be referentially different because
+        // they are either new mounts or alternates of previous children.
+        // Schedule updates and mounts depending on whether alternates exist.
+        // We don't track deletions here because they are reported separately.
+        if (nextChild.alternate) {
+          const prevChild = nextChild.alternate;
+          updateFiberRecursively(
+            nextChild,
+            prevChild,
+            shouldInclude ? nextFiber : parentFiber
+          );
+          // However we also keep track if the order of the children matches
+          // the previous order. They are always different referentially, but
+          // if the instances line up conceptually we'll want to know that.
+          if (!hasChildOrderChanged && prevChild !== prevChildAtSameIndex) {
+            hasChildOrderChanged = true;
           }
-          // Try the next child.
-          nextChild = nextChild.sibling;
-          // Advance the pointer in the previous list so that we can
-          // keep comparing if they line up.
-          if (!hasChildOrderChanged && prevChildAtSameIndex != null) {
-            prevChildAtSameIndex = prevChildAtSameIndex.sibling;
+        } else {
+          mountFiberRecursively(
+            nextChild,
+            shouldInclude ? nextFiber : parentFiber
+          );
+          if (!hasChildOrderChanged) {
+            hasChildOrderChanged = true;
           }
         }
-        // If we have no more children, but used to, they don't line up.
+        // Try the next child.
+        nextChild = nextChild.sibling;
+        // Advance the pointer in the previous list so that we can
+        // keep comparing if they line up.
         if (!hasChildOrderChanged && prevChildAtSameIndex != null) {
-          hasChildOrderChanged = true;
+          prevChildAtSameIndex = prevChildAtSameIndex.sibling;
         }
       }
-
-      if (shouldEnqueueUpdate) {
-        enqueueUpdateIfNecessary(nextFiber, hasChildOrderChanged);
+      // If we have no more children, but used to, they don't line up.
+      if (!hasChildOrderChanged && prevChildAtSameIndex != null) {
+        hasChildOrderChanged = true;
       }
     }
+
+    if (shouldInclude) {
+      maybeRecordUpdate(nextFiber, hasChildOrderChanged);
```

Comment:
nit the combination of `shouldInclude` and `maybeRecordUpdate` confused me at first glance. Maybe `shouldInclude` could be a little more obvious at a glance if it had a more detailed name? e.g. `shouldIncludeInTree`?